clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -439,25 +454,36 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         // Ensure signal handlers return to the interruption point.
         if (shard->prev_xfer_marker_.marker.marker_type ==
             TRACE_MARKER_TYPE_KERNEL_XFER) {
+            // We use the values popped from the signal-related stacks at the last
+            // TRACE_MARKER_TYPE_KERNEL_XFER marker.
             report_if_false(
                 shard,
-                ((memref.instr.addr == shard->prev_xfer_int_pc_.top() ||
-                  // DR hands us a different address for sysenter than the
-                  // resumption point.
-                  shard->pre_signal_instr_.top().instr.type ==
-                      TRACE_TYPE_INSTR_SYSENTER) &&
+                ((
+                     // Skip this check if we did not see the corresponding
+                     // kernel_event marker in the trace because the trace
+                     // started mid-signal.
+                     shard->last_xfer_int_pc_ == 0 ||
+                     memref.instr.addr == shard->last_xfer_int_pc_ ||
+                     // DR hands us a different address for sysenter than the
+                     // resumption point.
+                     shard->last_pre_signal_instr_.instr.type ==
+                         TRACE_TYPE_INSTR_SYSENTER) &&
                  (
-                     // Skip pre_signal_instr_ check if there was no such instr.
-                     shard->pre_signal_instr_.top().instr.addr == 0 ||
-                     memref.instr.addr == shard->pre_signal_instr_.top().instr.addr ||
+                     // Skip pre-signal instr check if there was no such instr.
```

Comment:
Should we split these into two different error strings: the post==marker (above) with one error string and the pre==post (below) with a different one?  May make it easier to analyze/group reported invariants, and split the complex boolean expression here up too.

Reply:
I split the boolean expression to simplify it.

But I think the different error messages are being worked on already in Sahil's PR. So I'll leave it to that.